### PR TITLE
8269725: AArch64: Add VectorMask query implementation for NEON

### PR DESCRIPTION
**Pull request by XiaohongGong**
The VectorMask query (`trueCount, firstTrue, lastTrue`) APIs can be intrinsified after [1] is closed. This patch adds the Arm NEON backend implementation for the new added vector nodes.

Here is the performance comparison data for the three APIs with and without this patch:
```
Benchmark                                        (bits) (inputs) Before       After      Gain  Units
MaskQueryOperationsBenchmark.testFirstTrueByte    128      1    42583.141   103900.253   2.44  ops/ms
MaskQueryOperationsBenchmark.testFirstTrueByte    128      2    37158.470   108234.110   2.91  ops/ms
MaskQueryOperationsBenchmark.testFirstTrueByte    128      3    42583.584   108235.231   2.54  ops/ms
MaskQueryOperationsBenchmark.testFirstTrueInt     128      1    42583.625   108236.859   2.54  ops/ms
MaskQueryOperationsBenchmark.testFirstTrueInt     128      2    42583.288   107368.205   2.52  ops/ms
MaskQueryOperationsBenchmark.testFirstTrueInt     128      3    42583.673   108232.371   2.54  ops/ms
MaskQueryOperationsBenchmark.testFirstTrueLong    128      1    42583.408   108232.617   2.54  ops/ms
MaskQueryOperationsBenchmark.testFirstTrueLong    128      2    42583.443   107367.035   2.52  ops/ms
MaskQueryOperationsBenchmark.testFirstTrueLong    128      3    42583.111   108236.036   2.54  ops/ms
MaskQueryOperationsBenchmark.testFirstTrueShort   128      1    42583.536   108230.365   2.54  ops/ms
MaskQueryOperationsBenchmark.testFirstTrueShort   128      2    41231.639   108239.148   2.62  ops/ms
MaskQueryOperationsBenchmark.testFirstTrueShort   128      3    42583.630   108238.542   2.54  ops/ms
MaskQueryOperationsBenchmark.testLastTrueByte     128      1    42584.067   108238.989   2.54  ops/ms
MaskQueryOperationsBenchmark.testLastTrueByte     128      2    36845.596   108234.297   2.94  ops/ms
MaskQueryOperationsBenchmark.testLastTrueByte     128      3    42583.759   108237.501   2.54  ops/ms
MaskQueryOperationsBenchmark.testLastTrueInt      128      1    42583.319   108236.218   2.54  ops/ms
MaskQueryOperationsBenchmark.testLastTrueInt      128      2    42583.112   108234.516   2.54  ops/ms
MaskQueryOperationsBenchmark.testLastTrueInt      128      3    42583.340   108238.777   2.54  ops/ms
MaskQueryOperationsBenchmark.testLastTrueLong     128      1    42581.004   108233.701   2.54  ops/ms
MaskQueryOperationsBenchmark.testLastTrueLong     128      2    42583.266   108238.323   2.54  ops/ms
MaskQueryOperationsBenchmark.testLastTrueLong     128      3    42583.542   108234.327   2.54  ops/ms
MaskQueryOperationsBenchmark.testLastTrueShort    128      1    42583.552   108238.011   2.54  ops/ms
MaskQueryOperationsBenchmark.testLastTrueShort    128      2    41231.142   108237.919   2.63  ops/ms
MaskQueryOperationsBenchmark.testLastTrueShort    128      3    44784.270   108238.011   2.42  ops/ms
MaskQueryOperationsBenchmark.testTrueCountByte    128      1    37075.556   108233.571   2.92  ops/ms
MaskQueryOperationsBenchmark.testTrueCountByte    128      2    37527.370   108233.396   2.88  ops/ms
MaskQueryOperationsBenchmark.testTrueCountByte    128      3    36585.788   107372.032   2.93  ops/ms
MaskQueryOperationsBenchmark.testTrueCountInt     128      1    42583.608   108233.721   2.54  ops/ms
MaskQueryOperationsBenchmark.testTrueCountInt     128      2    42584.733   107369.578   2.52  ops/ms
MaskQueryOperationsBenchmark.testTrueCountInt     128      3    42583.623   107367.859   2.52  ops/ms
MaskQueryOperationsBenchmark.testTrueCountLong    128      1    42583.671   107368.004   2.52  ops/ms
MaskQueryOperationsBenchmark.testTrueCountLong    128      2    42583.661   108233.301   2.54  ops/ms
MaskQueryOperationsBenchmark.testTrueCountLong    128      3    42583.015   108232.783   2.54  ops/ms
MaskQueryOperationsBenchmark.testTrueCountShort   128      1    41229.280   108233.369   2.63  ops/ms
MaskQueryOperationsBenchmark.testTrueCountShort   128      2    41231.914   107366.904   2.60  ops/ms
MaskQueryOperationsBenchmark.testTrueCountShort   128      3    41231.734   108233.606   2.63  ops/ms
```
All VectorAPI jtreg tests pass with patch [2] is applied together.

[1] https://bugs.openjdk.java.net/browse/JDK-8256973
[2] https://github.com/openjdk/jdk17/pull/168

Tested tier1 and jdk:tier3.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269725](https://bugs.openjdk.java.net/browse/JDK-8269725): AArch64: Add VectorMask query implementation for NEON

### Reviewers
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4699/head:pull/4699` \
`$ git checkout pull/4699`

Update a local copy of the PR: \
`$ git checkout pull/4699` \
`$ git pull https://git.openjdk.java.net/jdk pull/4699/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4699`

View PR using the GUI difftool: \
`$ git pr show -t 4699`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4699.diff">https://git.openjdk.java.net/jdk/pull/4699.diff</a>

</details>

### CODE REVIEW

**Comment by XiaohongGong** [comment]
Hi there, could anyone please take a look at this PR? Thanks so much!

File: src/hotspot/cpu/aarch64/aarch64_neon_ad.m4
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:2299]: Should that be "Reverse the bits?" But in any case, we can see that the code calls rbit then clz, presumably because you want to count the trailing bits. What does the negr do here?

Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:2299]: Thanks for your comment @theRealAph !  Yeah, it's "Reverse the bits".  Thanks for pointing out this!
`negr` is used to convert the original mask values from `0/1` to `0/-1`. So that we can firstly calculate the trailing zero numbers in bits, and then divide it by 8. Note that the mask input for these operations are different from others. It is actually a boolean vector with `0/1` as the element value.

Comment (theRealAph) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:2299]: Yes, but you reverse the bits in $dst, which reverses the byte order and changes each 0x01 byte to 0x80. You then count the leading zeroes. What, then, is the point of negating each 0x01 byte to 0xff first?

Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:2299]: Yes, I was intending to make the bits right after reversing. But seems it's also ok to remove the `negr` for `firsttrue`. And for `lasttrue`, I think it's better to keep the `negr`. I will remove it for `firsttrue` and have a test. Thanks for the review!

Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:2299]: Seems `negr` is also needless for `lasttrue`, since `7 >> 3 == 0` ?

Comment (nsjian) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:2299]: > Seems `negr` is also needless for `lasttrue`, since `7 >> 3 == 0` ?

Yeah. The in-reg mask representation of these VectorMaskOp nodes (0/1) are different from anytrue/alltrue (VectorTest, 0/-1), which looks a bit confusing.

Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4]: Hi @theRealAph , the new commit has removed the redundant `negr` for `firsttrue, lasttrue`, and all tests pass. Would you mind having look at it again? Thanks so much!

Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4]: > > Seems `negr` is also needless for `lasttrue`, since `7 >> 3 == 0` ?
> 
> Yeah. And the in-reg mask representation of these VectorMaskOp nodes (0/1) are different from anytrue/alltrue (VectorTest, 0/-1), which looks a bit confusing.

Yes, to make it clear, I'v added some comments for it. Thanks!

File: src/hotspot/cpu/aarch64/aarch64_neon.ad
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64_neon.ad:5340]: Say "vector of bytes." I read this as a vector of bits.

File: src/hotspot/cpu/aarch64/aarch64_neon.ad
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64_neon.ad:5381]: "64 bits". No hyphen.

File: src/hotspot/cpu/aarch64/aarch64_neon_ad.m4
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:2319]: These comments aren't very helpful. I can tell that this moves the lower 64-bits to a general register and checks whether the value is zero by reading the instructions. Your comments should say why, not how.
In reviewing this patch I had to search through the Java source to find out. It's "Returns the index of the first mask lane that is set in a mask of 16 bytes. Returns 16 (VLENGTH) if none of them are set."

File: src/hotspot/cpu/aarch64/aarch64_neon.ad
Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad:5340]: I use the "vector of boolean" here because the type of the input `(src)` is a `TypeVect` with `T_BOOLEAN` as the basic element type, although the size is the same with `byte`.

Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad:5340]: Sorry that I just got what you mean. So does "`vector of boolean with "0x00/0x01"`" make sense to you?

File: src/hotspot/cpu/aarch64/aarch64_neon_ad.m4
Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4]: Hi @theRealAph , I'v modified the comments in the latest commits. Could you please take a look at it again? Thanks so much!

File: src/hotspot/cpu/aarch64/aarch64_neon.ad
Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad]: Fixed, thanks!

**Comment by XiaohongGong** [comment]
Hi @theRealAph , all your significant comments have been addressed. Would you mind taking a look at this PR again? Thanks so much!

File: src/hotspot/cpu/aarch64/aarch64_neon.ad
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64_neon.ad:5355]: You should be able to use `cmpw($dst$$Register, vector_length(this, $src));` here if `operand_valid_for_add_sub_immediate(vector_length(this, $src))`

File: src/hotspot/cpu/aarch64/aarch64_neon_ad.m4
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:2318]: Why write `vmask_firsttrue8B` and `vmask_firsttrue_LT8B` separately? All you need is `if (vector_length < 8)` in the encoding rule.

**Review by theRealAph — COMMENTED** [review]
It's looking good, just a few minor issues.

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64.ad:1320]: Yes, actually they are not AAch64-specific. Besides AArch64, x86 also defines the same methods in `x86.ad` and they are used widely in there match rules.

File: src/hotspot/cpu/aarch64/aarch64_neon.ad
Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad:5355]: Good suggestion! I will fix this. Thanks so much!

File: src/hotspot/cpu/aarch64/aarch64_neon_ad.m4
Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:2318]: The main reason is the argument difference that `vmask_firsttrue_LT8B` needs to clear the `rFlagsReg` while `vmask_firsttrue8B`. It doesn't matter too much if the rule always clear the flag, right？

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64.ad:1320]: Since this change will touch the x86 codes, do you think it's better to create a followed-up patch to remove the shared methods?

File: src/hotspot/cpu/aarch64/aarch64_neon_ad.m4
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:2318]: It doesn't much matter, no, but being able to make differences like this without cut-and-paste programming is why we have a macro processor. Your call.

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/aarch64.ad:1320]: Yes, let's do it later. But let's not forget.

Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64.ad:1320]: Sure, will do it after this patch is merged. Thanks!

File: src/hotspot/cpu/aarch64/aarch64_neon.ad
Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad:5355]: @theRealAph , seems the `__ movw` in line-5354 is still needed since the `vector_length` is also used in the next instruction `__ cselw($dst$$Register, rscratch1, $dst$$Register, Assembler::GE);`. It is used to decide the final result.

File: src/hotspot/cpu/aarch64/aarch64.ad
Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64.ad:1320]: Recorded to a JBS here: https://bugs.openjdk.java.net/browse/JDK-8270519

File: src/hotspot/cpu/aarch64/aarch64_neon.ad
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad:5355]: So it is, you're right. Sorry.

Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_neon.ad:5355]: Never mind. I also forgot the last instruction yesterday! Thanks all the time for the suggestion!

File: src/hotspot/cpu/aarch64/aarch64_neon_ad.m4
Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:2318]: So for this part, do you mean it's ok if we can just merge the codes in m4 file while not in the ad file? Thanks!

Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4]: Hi @theRealAph , the latest commit has merged these two rules into a macro processor. Could you please to see whether it makes sense to you? Thanks so much!

**Review by theRealAph — APPROVED** [review]
OK, thanks.

**Comment by XiaohongGong** [comment]
Thanks for the review @theRealAph !

**Review by nsjian — APPROVED** [review]

**Pull request closed without merging**
